chain-impl-mockchain/src/testing/builders/vote.rs

```diff
@@ -48,13 +58,22 @@ pub fn decrypt_tally(
             let validated_tally = encrypted_tally
                 .validate_partial_decryptions(&members_pks, &decrypt_shares)
                 .expect("Invalid shares");
-            let tally = validated_tally.decrypt_tally(max_votes, &table).unwrap();
+
+            let tally = match (&table, max_votes.try_into()) {
+                (Some(table), Ok(max_votes)) => {
+                    validated_tally.decrypt_tally(table, max_votes).unwrap()
+                }
+                _ => Tally {
+                    votes: vec![0; validated_tally.len()],
+                },
+            };
```

Comment:
`decrypt_tally` is used in multiple modules across different repositories. We would have to repeat this code (and the above `TallyOptimizationTable` part) multiple times across different places. To avoid that I would suggest using an structure (let's call it `TallyDecryptor`) defined like that

```rust
struct TallyDecryptor {
    maybe_tally_optimization_table: Option<TallyOptimizationTable>
}
```

This structure should:

1. Initialize `TallyOptimizationTable` or set it to `None` in a conventional `new` method.
2. Expose `TallyDecryptor::decrypt_tally(&self, validated_tally: ValidatedTally, max_votes: u64)` that would encapsulate the code covered by this comment.

An example usage:

```rust
let tally_decryptor = TallyDecryptor::new(absolute_max_votes);
let tally = tally_decryptor.decrypt_tally(validated_tally, max_votes);
```

You can find examples of multiple usage in at least https://github.com/input-output-hk/jormungandr.

Reply:
Agree, I feel like currently the API is not at the correct level of abstraction, since exposing `TallyOptimizationTable` is not really relevant to downstream consumers, especially if it causes the problems outlined in this discussion.

Probably we could refine this even further, removing the need for passing `max_votes` and just exposing a function (or struct for a bit more flexibility) for decrypting a batch of `ValidatedTally`. 
E.g. something like:
```
fn batch_decrypt(tallies: &[ValidatedTally]) -> Result<Vec<Tally>> {
    let max_total_stake = tallies.iter().max_total_stake().max();
    if max_total_stake = 0 {
       ...
   }
    let table = ...
    tallies.iter()
             .map(|validate_tally| validated_tally.decrypt_tally(&table, validated_tally.max_total_stake()))
             .collect()
}
```
This would require us to keep track of the maximum total stake inside each `EncryptedTally/ValidatedTally`,  which is rather easy, we already do that but at another level https://github.com/input-output-hk/chain-libs/blob/2744ba2d3980e025b728293cb2ec08372d5f0526/chain-impl-mockchain/src/vote/tally.rs#L74). And we could also have a much more granular control on this value (by tracking https://github.com/input-output-hk/chain-libs/blob/2744ba2d3980e025b728293cb2ec08372d5f0526/chain-vote/src/tally.rs#L120) instead of using the total available stake https://github.com/input-output-hk/chain-libs/blob/2744ba2d3980e025b728293cb2ec08372d5f0526/chain-impl-mockchain/src/vote/manager.rs#L316.